src/lib/block.js

```diff
@@ -0,0 +1,136 @@
+'use strict'
+
+const {
+  UnixFS: { unmarshal: decodeUnixFs }
+} = require('ipfs-unixfs')
+const config = require('../config')
+const { now, queuedTasks } = require('./util')
+const { serializeError } = require('./logging')
+const { cidToKey, batchWriteDynamoItems } = require('./storage')
+const { publishBatch } = require('./publish')
+
+function blockType(blockId, codecs = config.codecs) {
+  return codecs[blockId.code]?.label ?? 'unsupported'
+}
+
+function blockData(block) {
+  if (!block.data) {
+    return
+  }
+
+  const data = decodeBlock(block)
+  if (data.Links) {
+    for (const link of data.Links) {
+      link.Hash = link.Hash?.toString()
+    }
+  }
+
+  return data
+}
+
+function decodeBlock(block, codecs = config.codecs) {
+  const codec = codecs[block.cid.code]
+
+  if (!codec) {
+    throw new Error(`Unsupported codec ${block.cid.code} in the block at offset ${block.blockOffset}`)
+  }
+
+  const data = codec.decode(block.data)
+
+  if (codec.label.startsWith('dag')) {
+    const { type, blocks } = decodeUnixFs(data.Data)
+
+    data.Data = { type, blocks }
+  }
+
+  return data
+}
+
+async function publishBlocks({ blocks, logger, queue = config.publishingQueue }) {
+  await publishBatch({ queue, messages: blocks.map(b => b.key), logger })
+}
+
+function storeBlocksTaskGenerator({ blocks, car, logger }) {
+  return async function storeBlocksTask() {
+    await writeBlocksBatch({ blocks, car, logger })
+    return { blocks, car }
+  }
+}
+
+async function storeBlocks({ car, source, logger, batchSize = config.blocksBatchSize, concurrency = config.concurrency, onTaskComplete }) {
+  let count = 0
+
+  const batch = []
+  const writes = queuedTasks({ concurrency, onTaskComplete })
+
+  try {
+    for await (const block of source) {
+      if (batch.length === batchSize) {
+        // the return of storeBlocksTask will be passed to publishBlocks
+        writes.add(storeBlocksTaskGenerator({ blocks: [...batch], car, logger }))
+        batch.length = 0
```

Comment:
Nit: personally I'd pass the batch and create a new array rather than copy the batch then truncate. For 2 reasons 1) IMO manipulating `length` to truncate is a lesser known feature of arrays (readability), and 2) I expect it's faster to create a new empty array than iterate over the existing one to create a copy.

```suggestion
  let batch = []
  const writes = queuedTasks({ concurrency, onTaskComplete })

  try {
    for await (const block of source) {
      if (batch.length === batchSize) {
        // the return of storeBlocksTask will be passed to publishBlocks
        writes.add(storeBlocksTaskGenerator({ blocks: batch, car, logger }))
        batch = []
```